clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -244,9 +255,50 @@ public void send(String method, Object... args) throws Exception {
         }
 
         InvocationMessage invocationMessage = new InvocationMessage(method, args);
-        String message = protocol.writeMessage(invocationMessage);
-        logger.log(LogLevel.Debug, "Sending message");
-        transport.send(message);
+        sendHubMessage(invocationMessage);
+    }
+
+    public <T> CompletableFuture<T> invoke(Class<T> returnType, String method, Object... args) throws Exception {
+        InvocationMessage invocationMessage = new InvocationMessage(method, args);
+        String id = connectionState.getNextId();
+        invocationMessage.setInvocationId(id);
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        connectionState.pendingInvocations.compute(id, (k, v) -> {
+            if (v != null) {
+                // This should never happen
+                throw new IllegalStateException("Invocation Id is already used");
+            }
+
+            InvocationRequest irq = new InvocationRequest(returnType);
+            CompletableFuture<Object> pendingCall = irq.getPendingCall();
+            // forward the invocation result to the user
+            pendingCall.thenAccept((returnValue) -> {
+                // Primitive types can't be cast with the Object cast function
+                if (returnType.isPrimitive()) {
+                    future.complete((T)returnValue);
+                } else {
+                    future.complete(returnType.cast(returnValue));
+                }
+            });
+
+            // forward any exceptions to the user
+            pendingCall.exceptionally((ex) -> future.completeExceptionally(ex));
+
+            return irq;
+        });
+
+        // Make sure the actual send is after setting up the future otherwise there is a race
+        // where the map doesn't have the future yet when the response is returned
+        sendHubMessage(invocationMessage);
+
+        return future;
+    }
+
+    private void sendHubMessage(HubMessage message) throws Exception {
+        String serializedMessage = protocol.writeMessage(message);
+        logger.log(LogLevel.Debug, "Sending message.");
```

Comment:
Put a little more detail in here maybe? What kind of message are you sending and the invocation ID if there is one? (see https://github.com/aspnet/SignalR/blob/release/2.2/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.Log.cs#L28)